### PR TITLE
feat(Select): add next version using menu components

### LINKED ISSUE #7321: Build simple ComposableSelect, move current Select examples to Legacy tab

### PR DESCRIPTION
**Pull request by kmcfaul**
**What**: Closes #7321

Adds select-next, a select component that utilizes the newer menu components.
Adds a handful of basic select demos to start with as well.

### CODE REVIEW

File: packages/react-core/src/next/components/Select/examples/SelectTypeahead.tsx
Comment (tlabaj) [inline packages/react-core/src/next/components/Select/examples/SelectTypeahead.tsx:2]: I wonder if we want this one to actually be a demo.

File: packages/react-core/src/next/components/Select/examples/SelectMultiTypeahead.tsx
Comment (tlabaj) [inline packages/react-core/src/next/components/Select/examples/SelectMultiTypeahead.tsx:2]: Same comment about maybe this should e a demo.

File: packages/react-core/src/next/components/Select/Select.tsx
Comment (tlabaj) [inline packages/react-core/src/next/components/Select/Select.tsx:28]: We should add ouia props to the select since the current version has them.  I opened an issue to do so for the next dropdown as well.

File: packages/react-core/src/next/components/Select/examples/Select.md
Comment (tlabaj) [inline packages/react-core/src/next/components/Select/examples/Select.md:11]: I would add an example for a Multi select since it s a very common use case.  An example with a grouped select would be useful as well.

**Review by tlabaj — CHANGES_REQUESTED** [review]
Looking great Katie!  I have a few comments.

File: packages/react-core/src/next/components/Select/SelectOption.tsx
Comment (thatblindgeye) [inline packages/react-core/src/next/components/Select/SelectOption.tsx:14]: Nit: just to follow the other boolean props, we could do "Flag indicating the option is disabled" instead.

File: packages/react-core/src/next/components/Select/SelectOption.tsx
Comment (thatblindgeye) [inline packages/react-core/src/next/components/Select/SelectOption.tsx:13]: Nitpick which would really depend if we'd want to make the change across all other components at some point, but this prop *could* alternatively be `hasCheckbox`.

File: packages/react-core/src/next/components/Select/SelectOption.tsx
Comment (thatblindgeye) [inline packages/react-core/src/next/components/Select/SelectOption.tsx:25]: Should the rest of the props be included here instead of just the spread props?

File: packages/react-core/src/next/components/Select/examples/SelectCheckbox.tsx
Comment (thatblindgeye) [inline packages/react-core/src/next/components/Select/examples/SelectCheckbox.tsx:14]:
```suggestion
        const firstElement = menuRef.current.querySelector('li input:not(:disabled)');
```

**Review by thatblindgeye — CHANGES_REQUESTED** [review]
In addition to the comments below, if it's something that could be added into this PR it'd be worth adding some logic to place focus back onto the toggle when an item is selected. Currently selecting an item doesn't provide an visual focus.

File: packages/react-core/src/next/components/Select/SelectOption.tsx
Comment (kmcfaul) [reply packages/react-core/src/next/components/Select/SelectOption.tsx:13]: Hmm yeah I can see that making more sense. Is it worth switching to `hasCheckbox` here right now or is it better to remain consistent for now and update the props in a breaking change release?

Comment (thatblindgeye) [reply packages/react-core/src/next/components/Select/SelectOption.tsx:13]: Good point. I'd be fine with keeping the prop as it is for that consistency  and doing a mass update in a breaking change release.

File: packages/react-core/src/next/components/Select/SelectOption.tsx
Comment (kmcfaul) [reply packages/react-core/src/next/components/Select/SelectOption.tsx:25]: They could but they would just be remapped to matching prop names on the Menu component. Spreading is just shorthanding `itemId={itemId} hasCheck={hasCheck} isDisabled={isDisabled}`, etc. LMK if we want to list the out instead.

On a side note about props, I don't list all of the Menu props, only the ones I used to make the current examples. But technically all Menu props will work because these extend the Menu props. I'm not sure if there's a clean way to indicate that though beyond looking at the Select prop interface.

File: packages/react-core/src/next/components/Select/examples/SelectTypeahead.tsx
Comment (kmcfaul) [reply packages/react-core/src/next/components/Select/examples/SelectTypeahead.tsx:2]: I can see it either way, though lean slightly closer to example because typeahead/multi-typeahead are pretty standard select use cases. 

All these examples use other components (MenuToggle and what goes into the toggle) so it's a bit unclear whether that makes them count more as demos.

File: packages/react-core/src/next/components/Select/SelectOption.tsx
Comment (thatblindgeye) [reply packages/react-core/src/next/components/Select/SelectOption.tsx:25]: I don't think I have a strong opinion either way. I was more curious since the other components had listed other props vs spreading.

File: packages/react-core/src/next/components/Select/Select.tsx
Comment (kmcfaul) [reply packages/react-core/src/next/components/Select/Select.tsx:53]: Yeah I tried to add support for space in, but it seems I'm running into the same issue you described.

I also removed the extra handling from the toggle clicks. I had forgotten that had moved internally when I went back to create the examples.

Comment (thatblindgeye) [reply packages/react-core/src/next/components/Select/Select.tsx:53]: Yeah I imagine it might take a little more investigating and I don't think it's necessarily worth holding this PR up over (especially since this is beta anyway)

File: packages/react-core/src/next/components/Select/Select.tsx
Comment (kmcfaul) [reply packages/react-core/src/next/components/Select/Select.tsx:28]: Added, but if you can double check after the update that would be great. I haven't touched the OUIA much so want to make sure I added it to the correct place.

File: packages/react-core/src/next/components/Select/Select.tsx
Comment (kmcfaul) [reply packages/react-core/src/next/components/Select/Select.tsx:53]: I'll open a follow up to investigate space

File: packages/react-core/src/next/components/Select/SelectOption.tsx
Comment (kmcfaul) [reply packages/react-core/src/next/components/Select/SelectOption.tsx:13]: Opened #8126 to track that potential breaking change

**Review by jenny-s51 — COMMENTED** [review]
These Select updates are so exciting @kmcfaul !

Just a quick question - I'm noticing this on a few of the new `react-next` demos... some of them have the Select menu width constrained to 200px rather than taking up the full width of their container! Was this intentional?

**Comment by kmcfaul** [comment]
Yes, the non typeahead selects would try to fit the toggle content otherwise, which would be too narrow for the menu when opened and also would change size when something got selected (for the selects that update the toggle text).

File: packages/react-core/src/next/components/Select/examples/SelectTypeahead.tsx
Comment (tlabaj) [reply packages/react-core/src/next/components/Select/examples/SelectTypeahead.tsx:2]: yeah... I was debating even commenting....

File: packages/react-core/src/next/components/Select/Select.tsx
Comment (tlabaj) [inline packages/react-core/src/next/components/Select/Select.tsx:102]: looks good.  Can you are the me to the props table.

**Comment by kmcfaul** [comment]
Yeah I think that's a good idea. I'll add that to the follow up

File: packages/react-core/src/next/components/Select/examples/SelectTypeahead.tsx
Comment (kmcfaul) [reply packages/react-core/src/next/components/Select/examples/SelectTypeahead.tsx:2]: Yeah it's a tricky set of examples for sure. I think we can probably leave the basic use cases as examples for now at least, and then when we get into the more complex use cases (like favorites, extra actions, etc) maybe those may be more fitting for demos.

**Review by mcarrano — DISMISSED** [review]
Can we make the typeahead examples more similar to the examples in the standard Select using a list of states for the options? I'm also noticing that these work differently with regards to what happens when you select a value. The old type ahead clears the field after selecting a value and restores the placeholder text. the Next version keeps the partial string that I typed in the field. I feel like the old behavior is what I expect. @mmenestr do you agree?

Also, the "Multiple single" example does not exist on the original Select and I question where this would be used vs the checkbox select. Don't they do the same thing?

**Comment by mmenestr** [comment]
This looks correct to me! The only thing I wonder is whether we should show a badge inside the toggle on the selects where a user can select more than one item to match the regular React demos

**Comment by kmcfaul** [comment]
@mcarrano @mmenestr 
The composable menu examples include a multiple select that is styled after the single select. It is the same in purpose compared to the checkbox select, just a different style that also is supported. LMK if you think we should remove this example.

Updated the typeahead examples to have the state data set, the multiple selects to show a badge when there are selections, and also updated the typeahead behavior to better match the original select example.

**Comment by mcarrano** [comment]
> The composable menu examples include a multiple select that is styled after the single select. It is the same in purpose compared to the checkbox select, just a different style that also is supported. LMK if you think we should remove this example.

@kmcfaul I'm looking at how this is presented in the design guidelines and there is a reference to a multi-select as in the Multiple-single example. However this suggests use of the chips as in a typeahead to differentiate it from the checkbox select. Not sure we ever exactly had a code example for that. Can we create that example using the new approach? @mmenestr interested to hear your perspective. How should we better align the React examples with our design guidance?

**Comment by tlabaj** [comment]
@mcarrano @mmenestr @kmcfaul When I suggested adding a multi select.  I thought it would match the example in the current select component where we have chips of the selected items. Here https://patternfly-react-pr-8115.surge.sh/components/select#multiple

**Comment by mmenestr** [comment]
@mcarrano @tlabaj agree! I think we should get rid of the multiple single example because when we have a multiple select with the checkmarks, i'd except the chips to always show up in the toggle to match the multiple select examples we have in our design guidance.

File: packages/react-core/src/next/components/Select/examples/SelectGrouped.tsx
Comment (mcoker) [inline packages/react-core/src/next/components/Select/examples/SelectGrouped.tsx:58]: `<SelectList>` is a `<ul>` so its children need to be `<SelectOption>` since those are `<li>`'s

```suggestion
      <SelectGroup label="Group 1">
        <SelectList>
          <SelectOption itemId="Option 1">Option 1</SelectOption>
          <SelectOption itemId="Option 2">Option 2</SelectOption>
          <SelectOption itemId="Option 3">Option 3</SelectOption>
        </SelectList>
      <SelectGroup label="Group 2">
        <SelectList>
          <SelectOption itemId="Option 4">Option 4</SelectOption>
          <SelectOption itemId="Option 5">Option 5</SelectOption>
          <SelectOption itemId="Option 6">Option 6</SelectOption>
        </SelectList>
      </SelectGroup>
```

File: packages/react-core/src/next/components/Select/examples/Select.md
Comment (mcoker) [inline packages/react-core/src/next/components/Select/examples/Select.md:24]: Should this just be "multiple"?

```suggestion
### Multiple
```

File: packages/react-core/src/next/components/Select/examples/SelectTypeahead.tsx
Comment (mcoker) [inline packages/react-core/src/next/components/Select/examples/SelectTypeahead.tsx:20]: Hmm... for some reason, I'm not seeing these items in the list (or the `placeholder`) for this or the multi typeahead demo. Maybe that's an issue with surge on my end?

**Review by mcoker — CHANGES_REQUESTED** [review]

File: packages/react-core/src/next/components/Select/SelectGroup.tsx
Comment (kmcfaul) [reply packages/react-core/src/next/components/Select/SelectGroup.tsx:20]: Yes it is possible, as all of these select components extend the menu components. In this case the `labelHeadingLevel` prop from MenuGroup would be used to set the heading level.

Perhaps the larger question is how we want to organize these extended props. I'm not a huge fan of duplicating the whole prop list into each Select component in case anything gets updated in Menu later on. Maybe in the docs I can pull in the Menu props as well as the Select props? @tlabaj

File: packages/react-core/src/next/components/Select/SelectOption.tsx
Comment (kmcfaul) [reply packages/react-core/src/next/components/Select/SelectOption.tsx:19]: Yep for the same reason as above

File: packages/react-core/src/next/components/Select/examples/Select.md
Comment (kmcfaul) [reply packages/react-core/src/next/components/Select/examples/Select.md:24]: This example is being removed

File: packages/react-core/src/next/components/Select/examples/SelectTypeahead.tsx
Comment (kmcfaul) [reply packages/react-core/src/next/components/Select/examples/SelectTypeahead.tsx:20]: That's odd I am also not seeing these on the preview (the code is a commit behind when I looked). Maybe with the next commit it will update the surge properly.

File: packages/react-core/src/next/components/Select/SelectGroup.tsx
Comment (kmcfaul) [reply packages/react-core/src/next/components/Select/SelectGroup.tsx:20]: Trying locally, it doesn't find the Menu components during tsDocgen because the next components are in a separate folder (so it's trying to parse the Menu props as menu-next files).

Comment (tlabaj) [reply packages/react-core/src/next/components/Select/SelectGroup.tsx:20]: @kmcfaul maybe we should open an issue in org so we can pull in props from src directory into next.  
AS far as exposing props for Menu.  I think if it is a common use case we can show it in the Select's prop table.  If it is something we are not really encouraging/ is not a common use case I would vote to not add it.

**Review by mcoker — APPROVED** [review]
LGTM!

One thing I noticed is we're generating an empty `id` on menu group titles if `titleId` (optional) isn't set - is that something we should fix? I can open another issue if so https://github.com/patternfly/patternfly-react/blob/af40c12e9dc55ed508ebc5e07415ec93d61bd639/packages/react-core/src/components/Menu/MenuGroup.tsx#L31

File: packages/react-core/src/next/components/Select/SelectGroup.tsx
Comment (kmcfaul) [reply packages/react-core/src/next/components/Select/SelectGroup.tsx:20]: I can open a quick follow up to go over the Menu props and see which ones we want to include for the dropdown/select wrappers & add them

**Comment by kmcfaul** [comment]
Let's open a follow up for the MenuGroup titleId if you agree @tlabaj . I can open a separate PR just to not keep this one held up.

**Review by tlabaj — APPROVED** [review]
LGTM as long as we open the follow up issue.

**Review by mmenestr — APPROVED** [review]

**Comment by kmcfaul** [comment]
Opened #8162 for the group id issue

**Review by jenny-s51 — APPROVED** [review]

**Pull request merged**
